security/s2a/internal/crypter/aesgcm_test.go

```diff
@@ -0,0 +1,474 @@
+	}
+	if string(decryptedPlaintext) != plaintext {
+		t.Fatalf("Decrypt(%v, %v, %v, nil) = %v, want = %v", ciphertext[:0], ciphertext, nonce, decryptedPlaintext, plaintext)
+	}
+}
+
+// Test encrypt and decrypt using an invalid key size.
+func TestAESGCMInvalidKeySize(t *testing.T) {
+	// Use 17 bytes, which is invalid
+	key := make([]byte, 17)
+	if _, err := NewAESGCM(key); err == nil {
+		t.Error("expected an error when using invalid key size")
+	}
+}
+
+// Test update key for AES-GCM using a key with different size from the initial
+// key.
+func TestAESGCMKeySizeUpdate(t *testing.T) {
+	for _, tc := range []struct {
+		desc          string
+		updateKeySize int
+	}{
+		{"mismatch key size update", aes256GcmKeySize},
+		{"invalid key size update", 17},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			key := make([]byte, aes128GcmKeySize)
+			crypter, err := NewAESGCM(key)
+			if err != nil {
+				t.Fatalf("NewAESGCM(keySize=%v) failed, err: %v", aes128GcmKeySize, err)
+			}
+
+			// Update the key with a new one which is a different from the original.
+			newKey := make([]byte, tc.updateKeySize)
+			if err = crypter.UpdateKey(newKey); err == nil {
+				t.Fatal("UpdateKey should fail with invalid key size error")
+			}
+		})
+	}
+}
+
+// Test Encrypt/Decrypt using an invalid nonce size.
+func TestAESGCMEncryptDecryptInvalidNonce(t *testing.T) {
+	key := make([]byte, aes128GcmKeySize)
+	crypter, err := NewAESGCM(key)
+	if err != nil {
+		t.Fatalf("NewAESGCM(keySize=%v) failed, err: %v", aes128GcmKeySize, err)
+	}
+	// Construct nonce with invalid size.
+	nonce := make([]byte, 1)
+	if _, err = crypter.Encrypt(nil, nil, nonce, nil); err == nil {
+		t.Errorf("Encrypt should fail due to invalid nonce size")
+	}
+	if _, err = crypter.Decrypt(nil, nil, nonce, nil); err == nil {
+		t.Fatalf("Decrypt should fail due to invalid nonce size")
+	}
+}
+
+// Test encrypt and decrypt on roundtrip messages for AES-GCM with and without
+// updating the keys.
+func TestAESGCMEncryptRoundtrip(t *testing.T) {
+	for _, shouldUpdateKeys := range []bool{true, false} {
+		for _, keySize := range []int{aes128GcmKeySize, aes256GcmKeySize} {
+			key := make([]byte, keySize)
+			sender, receiver := getGCMCryptoPair(key, t)
+			if shouldUpdateKeys {
+				// Update the key with a new one which is different from the original.
+				newKey := make([]byte, keySize)
+				newKey[0] = '\xbd'
+				if err := sender.UpdateKey(newKey); err != nil {
+					t.Fatalf("sender UpdateKey failed with: %v", err)
+				}
+				if err := receiver.UpdateKey(newKey); err != nil {
+					t.Fatalf("receiver UpdateKey failed with: %v", err)
+				}
+			}
+			testGCMEncryptRoundtrip(sender, receiver, t)
+		}
+	}
+}
+
+// Test encrypt and decrypt using test vectors for aes128gcm. Much of these
+// test vectors were taken from the ALTS AES-GCM unit tests.
+func TestAESGCMEncrypt(t *testing.T) {
+	for _, test := range []testutil.CryptoTestVector{
```

Comment:
Please use the `Comment` field here as well and pass it to `t.Run`. See my comment below about that.